### PR TITLE
Prevent loops in extraction based on merge

### PR DESCRIPTION
**Pull request by MoKob**
# Issue

https://github.com/Project-OSRM/osrm-backend/issues/2927. Turns out that we can end up in loops of degree two due to merging of segregated roads. This requires a strange set of modelled roads, but we can't  assure it will not be in the data.

This PR implements a fallback that assures we cannot loop over traffic-lights endlessly.

## Tasklist
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [ ] review
 - [ ] adjust for for comments

### CODE REVIEW

File: features/guidance/dedicated-turn-roads.feature [hunk 1]
```diff
@@ -313,3 +313,39 @@ Feature: Slipways and Dedicated Turn Lanes
...
+            | nodes | highway | name | oneway |
+            | abi   | primary | road | yes    |
+            | bcjd  | primary | loop | yes    |
+            | dhgf  | primary | loop | yes    |
+            | fed   | primary | loop | yes    |
+
+        And the nodes
+            | node | highway         |
+            | g    | traffic_signals |
+            | c    | traffic_signals |
```
Comment (daniel-j-h) [inline features/guidance/dedicated-turn-roads.feature]: Are these required to trigger the endless looping?

File: features/guidance/roundabout.feature [hunk 2]
```diff
@@ -36,6 +36,25 @@ Feature: Basic Roundabout
...
+    #2927
+    Scenario: Only Roundabout
+        Given the node map
+            |   |   | a |   |   |
+            |   |   |   |   |   |
+            |   |   |   |   |   |
+            | b |   |   |   | d |
+            |   |   |   |   |   |
+            |   |   |   |   |   |
+            |   |   | c |   |   |
```
Comment (daniel-j-h) [inline features/guidance/roundabout.feature]: Is this related to the issue? I thought the loop was based on merges?

File: src/extractor/guidance/intersection_generator.cpp [hunk 3]
```diff
@@ -577,9 +580,13 @@ IntersectionGenerator::GetActualNextIntersection(const NodeID starting_node,
...
-    while (result.size() == 2 &&
-           node_based_graph.GetEdgeData(via_edge).IsCompatibleTo(
-               node_based_graph.GetEdgeData(result[1].turn.eid)))
+    // using a maximum lookahead, we make sure not to end up in some form of loop
+    const constexpr std::size_t max_lookahead = 10;
+    for (std::size_t lookahead = 0;
+         lookahead < max_lookahead && (result.size() == 2 &&
+                                       node_based_graph.GetEdgeData(via_edge).IsCompatibleTo(
+                                           node_based_graph.GetEdgeData(result[1].turn.eid)));
+         ++lookahead)
```
Comment (daniel-j-h) [inline src/extractor/guidance/intersection_generator.cpp]: For my own understanding: you try to find the first compatible edge but now with a limit so that in edge cases (as the one triggering this) we only end up looking at a limited amount (10 here) to prevent loops. Why the 10 here instead of e.g. storing and checking for a actual loop?

File: src/extractor/guidance/sliproad_handler.cpp [hunk 4]
```diff
@@ -51,27 +51,36 @@ operator()(const NodeID, const EdgeID source_edge_id, Intersection intersection)
...
-        [&](const NodeID at_node, const ConnectedRoad &road, NodeID &output_node) {
-            auto intersection = intersection_generator(at_node, road.turn.eid);
-            auto in_edge = road.turn.eid;
-            // skip over traffic lights
-            while (intersection.size() == 2)
+    const auto findNextIntersectionForRoad = [&](
+        const NodeID at_node, const ConnectedRoad &road, NodeID &output_node) {
+        auto intersection = intersection_generator(at_node, road.turn.eid);
+        auto in_edge = road.turn.eid;
+        // skip over traffic lights
```
Comment (daniel-j-h) [inline src/extractor/guidance/sliproad_handler.cpp]: Ah this is the reason you had traffic lights in the test case at the top?

File: src/extractor/guidance/sliproad_handler.cpp [hunk 5]
```diff
@@ -51,27 +51,36 @@ operator()(const NodeID, const EdgeID source_edge_id, Intersection intersection)
...
+        // if there are more than 10 traffic lights/other items on the road, we don't consider it
+        // for sliproads anymore. This also prevents us from ending up in an endless loop.
```
Comment (daniel-j-h) [inline src/extractor/guidance/sliproad_handler.cpp]: This is primarily for endless loops, right? We don't really care for number of traffic lights on sliproads?

File: features/guidance/dedicated-turn-roads.feature (hunk 1 above)
Comment (MoKob) [inline features/guidance/dedicated-turn-roads.feature]: Yes, as you found out already.

File: features/guidance/roundabout.feature (hunk 2 above)
Comment (MoKob) [inline features/guidance/roundabout.feature]: This is simply an additional case where loops could theoretically happen (roundabout handler has loop checking as well). This is just additional drive-by security

File: src/extractor/guidance/intersection_generator.cpp (hunk 3 above)
Comment (MoKob) [inline src/extractor/guidance/intersection_generator.cpp]: Well, we could remember all nodes we see adding an additional data structure (e.g. hashmap) and check whether any node is repeated. Since we usually have a single traffic light to look at, this check is a simple way around broken modelings. At the same time, we don't have to allocate a dedicated `unordered_set` for every iteration.

If you feel that we should be on the save side for even situations with more than 10 traffic-lights/reasons not to compress on a single edge, we can go for an unordered_map, though.

File: src/extractor/guidance/sliproad_handler.cpp (hunk 5 above)
Comment (MoKob) [inline src/extractor/guidance/sliproad_handler.cpp]: No. We simply want to prevent endless loops. The same argument as above holds here. If we want to detect endless loops here, we would need an `unordered_set` to detect these. There is simply no situation that justifies a detection as sliproad if there is like 10 or more traffic lights in between. Anything but `1` would be unexpected.
